# Segmenting and Clustering Neighborhoods in Toronto

In [1]:
from bs4 import BeautifulSoup
import urllib.request
import pandas as pd

#### *Use BeautifulSoup to get the information from the website*

In [2]:
#Get the HTML file from the URL
html = urllib.request.urlopen('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').read()

In [3]:
soup = BeautifulSoup(html)

In [4]:
#Find the table in the HTML file and extract it into a dataframe
table = soup.find('table', attrs={'class':'wikitable sortable'})
table_rows = table.find_all('tr')
res = []
for tr in table_rows:
    td = tr.find_all('td')
    row = [tr.text.replace("\n", "") for tr in td]
    if row:
        res.append(row)
df = pd.DataFrame(res, columns=["Postal Code", "Borough", "Neighborhood"])
df

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                          Neighborhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

In [5]:
#Ignore cells with a borough that is Not assigned.
df = df[df.Borough != 'Not assigned']
df.shape

(103, 3)

In [6]:
df['Postal Code'].unique

<bound method Series.unique of 2      M3A
3      M4A
4      M5A
5      M6A
6      M7A
      ... 
160    M8X
165    M4Y
168    M7Y
169    M8Y
178    M8Z
Name: Postal Code, Length: 103, dtype: object>

#### *We have 103 rows in the dataframe (after the ignoration of Not assigned Borough).* 
#### *And we have 103 different postal code in the dataframe. See the unique() result.*
#### *Wiki may redo this table and the problem doesn't exist anymore.*

*Combine those neighborhoods who share the same postal caode into one row and separate them with a comma.*  
*If you want to run the combination for sure, uncomment the code below.*

In [7]:
#df.groupby('Postal Code').agg({'Borough' : 'first', 'Neighborhood' : ','.join}).reset_index().reindex(columns=df.columns)

In [8]:
#Find if the column of Neighborhood contain the string "Not assigned"
a = df.Neighborhood
if 'Not assigned' in a:
    print("The column of Neighborhood contains string 'Not assigned'")
else:
    print('The column of Neighborhood is good')

The column of Neighborhood is good


In [9]:
df = df.reset_index(drop = True)
print (df.shape)
df

(103, 3)


Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]